java/org/apache/tomcat/util/http/fileupload/ProgressListenerFactory.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.tomcat.util.http.fileupload;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+
+import org.apache.juli.logging.Log;
+import org.apache.tomcat.util.res.StringManager;
+
+/**
+ * Factory for {@link ProgressListener}. Users can specify this via a System property. See FACTORY_NAME
+ */
+public interface ProgressListenerFactory {
+
+    /**
+     * Constant for a system property that can be used to specify a factory for ProgressListeners
+     */
+    String FACTORY_NAME = "org.apache.tomcat.util.http.fileupload.ProgressListenerFactory";
+
+    /**
+     * Factory that creates no listeners. To avoid null checks in code.
+     */
+    ProgressListenerFactory NULL_FACTORY = new ProgressListenerFactory() {
+        @Override
+        public ProgressListener newProgressListener() {
+            return null;
+        }
+    };
+
+    static ProgressListenerFactory newFactory(final Log log, StringManager sm) {
```

Comment:
oK. I already see this design has the problem that different applications might live in one server. Thus, we need some way to make this application or Servlet context specific.

Reply:
But for my local proof of concept this is enough.